base/linalg/bidiag.jl

```diff
@@ -270,15 +270,13 @@ function triu!(M::Bidiagonal, k::Integer=0)
     return M
 end
 
-function diag(M::Bidiagonal{T}, n::Integer=0) where T
+function diag(M::Bidiagonal, n::Integer=0)
     if n == 0
-        return M.dv
-    elseif n == 1
-        return M.uplo == 'U' ? M.ev : zeros(T, size(M,1)-1)
-    elseif n == -1
-        return M.uplo == 'L' ? M.ev : zeros(T, size(M,1)-1)
+        return copy!(similar(M.dv, length(M.dv)), M.dv)
```

Comment:
Could the `length(M.dv)` be omitted?

Reply:
Moreover, why not simply `copy(...` rather than `copy!(similar(M.dv, length(M.dv)), M.dv)`? Is the return type dependent upon `k` in some cases (e.g. for ranges) otherwise?

Reply:
> Guarding against the length specification modifying the return type?

Yes, that was the idea, to guarantee that we take the same path no matter which diagonal was requested.

> Moreover, why not simply copy(... rather than copy!(similar(M.dv, length(M.dv)), M.dv)? Does the return type otherwise depend upon k in some cases (e.g. for ranges)?

Same comment applies; guarantee the same code path. Also `copy(r::AbstractRange) = r` so that will not yield a `Vector{Int}` as `similar` does.
```julia
julia> copy(1:4)
1:4
```

I don't think the `copy!(similar(...), ...)` uses are that terrible -- it is pretty clear what is going on IMO.

Reply:
Sounds good then! I might consider adding comments to make the intention clear for future readers; to someone not acquainted with the subtleties involved, the intention may be far from clear :).

Reply:
Yea, Ill add that